### PR TITLE
[Flax] Add flax marian

### PR DESCRIPTION
**Pull request by patrickvonplaten**
This PR adds Flax Marian

### CODE REVIEW

File: src/transformers/models/marian/modeling_flax_marian.py [hunk 1]
```diff
@@ -0,0 +1,1452 @@
...
+        )
+
+        def _decoder_forward(module, decoder_input_ids, decoder_attention_mask, decoder_position_ids, **kwargs):
+            decoder_module = module._get_decoder_module()
+            return decoder_module(
+                decoder_input_ids,
+                decoder_attention_mask,
+                decoder_position_ids,
+                **kwargs,
+            )
```
Comment (sgugger) [inline src/transformers/models/marian/modeling_flax_marian.py:945]:
```suggestion
            return decoder_module(decoder_input_ids, decoder_attention_mask, decoder_position_ids, **kwargs)
```
Fits in one line.

**Review by sgugger — APPROVED** [review]
Thanks for adding this!

File: src/transformers/models/marian/modeling_flax_marian.py [hunk 2]
```diff
@@ -0,0 +1,1453 @@
...
+            "encoder_outputs": encoder_outputs,
+            "encoder_attention_mask": attention_mask,
+            "decoder_attention_mask": extended_attention_mask,
+            "decoder_position_ids": position_ids,
+        }
+
+    def update_inputs_for_generation(self, model_outputs, model_kwargs):
+        model_kwargs["past_key_values"] = model_outputs.past_key_values
+        model_kwargs["decoder_position_ids"] = model_kwargs["decoder_position_ids"][:, -1:] + 1
+        return model_kwargs
```
Comment (patil-suraj) [inline src/transformers/models/marian/modeling_flax_marian.py:1453]: Not sure if we will have a similar problem here, but in `MarianMTModel` we have the `adjust_logits_during_generation` method which sets pad token logits `-inf` to never predict the pad.

https://github.com/huggingface/transformers/blob/8fe836af5a164dac60b8c45cc73a7b9ed72393a3/src/transformers/models/marian/modeling_marian.py#L1346-L1348

**Review by patil-suraj — APPROVED** [review]
Thanks a lot for adding this!

File: src/transformers/models/marian/modeling_flax_marian.py (hunk 2 above)
Comment (patrickvonplaten) [reply src/transformers/models/marian/modeling_flax_marian.py:1453]: Yeah I thought we can remove this hack, but too many languages seem to depend on it - just checked the results. So, we'll have to add it for Marian as well -> will adapt.

File: src/transformers/generation_flax_utils.py [hunk 3]
```diff
@@ -696,6 +704,9 @@ def beam_search_body_fn(state):
                 lambda tensor: unflatten_beam_dim(tensor, batch_size, num_beams), model_outputs.past_key_values
             )
 
+            # adapt logits for FlaxMarianMTModel
+            logits = self._adapt_logits_for_beam_search(logits)
```
Comment (patrickvonplaten) [inline src/transformers/generation_flax_utils.py:708]: @gchhablani - note that this should be added to generation for Marian
